Scenarios/Authentication/Mixed-auth-allow-anonymous-workaround/Idp/Program.cs

```diff
@@ -0,0 +1,74 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Security.Claims;
+using Microsoft.AspNetCore;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.IdentityModel.Tokens;
+using OpenIddict.Abstractions;
+using OpenIddict.Server;
+using OpenIddict.Server.AspNetCore;
+
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.AddOpenIddict()
+    .AddServer(options =>
+    {
+        options.AddDevelopmentEncryptionCertificate();
+        options.AddDevelopmentSigningCertificate();
+        options.AllowClientCredentialsFlow();
+        options.SetTokenEndpointUris("/token");
+        options.EnableDegradedMode();
+        options.UseAspNetCore();
+        options.RegisterScopes("api");
+        options.DisableAccessTokenEncryption();
+        options.AddEventHandler<OpenIddictServerEvents.ValidateTokenRequestContext>(options =>
+            options.UseInlineHandler(async context =>
+            {
+                if (!context.Request.IsClientCredentialsGrantType())
+                {
+                    context.Reject(error: OpenIddictConstants.Errors.InvalidGrant);
+                    return;
+                }
+
+                if (!string.Equals(context.ClientId, "console_app", StringComparison.Ordinal))
+                {
+                    context.Reject(error: OpenIddictConstants.Errors.InvalidClient);
+                    return;
+                }
+
+                if (!string.Equals(context.ClientSecret, "secret", StringComparison.Ordinal))
+                {
+                    context.Reject(error: OpenIddictConstants.Errors.InvalidClient);
+                    return;
+                }
+
+                if (string.IsNullOrEmpty(context.Request.Scope))
+                {
+                    context.Reject(error: OpenIddictConstants.Errors.InvalidScope);
+                    return;
+                }
+
+                var identity = new ClaimsIdentity(
+                    authenticationType: TokenValidationParameters.DefaultAuthenticationType,
+                    nameType: OpenIddictConstants.Claims.Name,
+                    roleType: OpenIddictConstants.Claims.Role);
+
+                identity.SetScopes("api");
+                identity.SetAudiences("api");
+                identity.SetClaim(OpenIddictConstants.Claims.Subject, context.ClientId);
+
+                context.Principal = new ClaimsPrincipal(identity);
+
+                var request = context.Transaction.GetHttpRequest();
+                await request!.HttpContext.SignInAsync(OpenIddictServerAspNetCoreDefaults.AuthenticationScheme, context.Principal);
```

Comment:
This is not the supported/recommended way to implement that. Instead, move the code after line 52 to a `HandleTokenRequest(Context)` event handler and replace the `SignInAsync()` + `context.HandleRequest()` call by `context.SetPrincipal(principal)`.

Reply:
I moved the code a new handler.